src/nvim/ex_getln.c

```diff
@@ -2306,6 +2313,270 @@ static int empty_pattern(char_u *p)
+  if (try_end(&err) || result == FAIL) {
+    api_clear_error(&err);
+    return NULL;
+  }
+
+  curwin->w_p_cul = false;
+  curwin->w_p_cuc = false;
+  curwin->w_p_spell = false;
+  curwin->w_p_fen = false;
+
+  win_enter(save_curwin, false);
+  return preview_win;
+}
+
+/// Closes any open windows for inccommand preview buffer.
+static void close_cmdpreview_win(void)
+{
+  buf_T *buf = cmdpreview_bufnr ? buflist_findnr(cmdpreview_bufnr) : NULL;
+  if (buf != NULL) {
+    close_windows(buf, false);
+  }
+}
+
+/// Show 'inccommand' preview. It works like this:
+///    1. Store current undo information so we can revert to current state later.
+///    2. Execute the preview callback with the parsed command, preview buffer number and preview
+///       namespace number as arguments. The preview callback sets the highlight and does the
+///       changes required for the preview if needed.
+///    3. Preview callback returns 0, 1 or 2. 0 means no preview is shown. 1 means preview is shown
+///       but preview window doesn't need to be opened. 2 means preview is shown and preview window
+///       needs to be opened if inccommand=split.
+///    4. Use the return value of the preview callback to determine whether to
+///       open the preview window or not and open preview window if needed.
+///    5. If the return value of the preview callback is not 0, update the screen while the effects
+///       of the preview are still in place.
+///    6. Revert all changes made by the preview callback.
+static void show_cmdpreview(CommandLineState *s)
+{
+  // Parse the command line and return if it fails.
+  exarg_T ea;
+  CmdParseInfo cmdinfo;
+  // Copy the command line so we can modify it.
+  char *cmdline = xstrdup((char *)ccline.cmdbuff);
+  char *errormsg = NULL;
+
+  parse_cmdline(cmdline, &ea, &cmdinfo, &errormsg);
+  if (errormsg != NULL) {
+    goto end;
+  }
+
+  // Swap invalid command range if needed
+  if ((ea.argt & EX_RANGE) && ea.line1 > ea.line2) {
+    linenr_T lnum = ea.line1;
+    ea.line1 = ea.line2;
+    ea.line2 = lnum;
+  }
+
+  time_t save_b_u_time_cur = curbuf->b_u_time_cur;
+  long save_b_u_seq_cur = curbuf->b_u_seq_cur;
+  u_header_T *save_b_u_newhead = curbuf->b_u_newhead;
+  long save_b_p_ul = curbuf->b_p_ul;
+  int save_b_changed = curbuf->b_changed;
+  bool save_no_send_tick = curbuf->no_send_tick;
+  int save_w_p_cul = curwin->w_p_cul;
+  int save_w_p_cuc = curwin->w_p_cuc;
+  bool save_hls = p_hls;
+  varnumber_T save_changedtick = buf_get_changedtick(curbuf);
+  buf_T *cmdpreview_buf;
+  win_T *cmdpreview_win;
+  cmdmod_T save_cmdmod = cmdmod;
+
+  cmdpreview = true;
+  emsg_silent++;                 // Block error reporting as the command may be incomplete
+  msg_silent++;                  // Block messages, namely ones that prompt
+  block_autocmds();              // Block events
+  curbuf->no_send_tick = true;   // Block sending b:changedtick in buffer updates
```

Comment:
can buffer-update logic make that decision based on the fact that the buffer is known to be the cmdpreview buffer?

there is a global state already...

Reply:
Good point actually, we can reuse that

Reply:
I've implemented it and removed the `no_send_tick` flag